### PR TITLE
[Feature] improve hash join performance by coroutine-based interleaving

### PR DESCRIPTION
**Pull request by fzhedu**
`interleaving_group_size` controls enable interleaving or not, default 10. It means how many coroutines to interleave per thread.
- 0 means disable interleaving;
- positive values mean adaptive interleaving with  group = `interleaving_group_size`, enable it if the hash table is enough large(>32MB);
- negtive values mean force interleaving with group = abs(`interleaving_group_size`);

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### performance improvement
image

```
--SQL_1 self join
select count(*) from lineitem t join lineitem s on s.l_partkey = t.l_orderkey;
--SQL_2 primary-foreign key join
select count(*) from orders join lineitem on o_orderkey = l_orderkey;
--SQL_3 big table join
select count(*) from store_sales join store_returns on ss_customer_sk = sr_customer_sk;
--SQL_4 sting join key
select count(*) from orders join lineitem on left(o_comment,64) = l_comment;
--SQL_5 multi join key
select count(*) from store_sales join store_returns on ss_customer_sk = sr_customer_sk and ss_ticket_number = sr_ticket_number;
```

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 281 / 345 (81.45%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/PartitionType.java|0|1|00.00%|[69]|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|0|5|00.00%|[942, 972, 973, 987, 1014]|
|:large_blue_circle:|com/starrocks/catalog/ExternalOlapTable.java|0|7|00.00%|[398, 437, 438, 554, 556, 557, 558]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/ColumnRefOperator.java|0|4|00.00%|[112, 113, 115, 117]|
|:large_blue_circle:|com/starrocks/qe/QueryState.java|2|3|66.67%|[192]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMvRefreshProcessor.java|107|131|81.68%|[218, 226, 378, 424, 584, 585, 590, 595, 596, 597, 598, 599, 601, 602, 603, 604, 606, 607, 611, 612, 903, 904, 905, 906]|
|:large_blue_circle:|com/starrocks/scheduler/MvTaskRunContext.java|9|11|81.82%|[104, 105]|
|:large_blue_circle:|com/starrocks/connector/PartitionUtil.java|47|58|81.03%|[288, 294, 424, 428, 432, 433, 434, 437, 443, 444, 484]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|19|22|86.36%|[96, 97, 98]|
|:large_blue_circle:|com/starrocks/scheduler/persist/MVTaskRunExtraMessage.java|7|8|87.50%|[98]|
|:large_blue_circle:|com/starrocks/authentication/AuthenticationMgr.java|48|53|90.57%|[253, 269, 270, 271, 289]|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/ColumnDict.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/SecurityIntegrationFactory.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/Statistics.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/system/SystemInfoService.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/ColumnStatistic.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MvUtils.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/ExpressionStatisticCalculator.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/privilege/AuthorizationMgr.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/authentication/UserAuthenticationInfo.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/ColumnBasicStatsCacheLoader.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/system/HeartbeatMgr.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

File: be/src/exec/hash_joiner.cpp [hunk 1]
```diff
@@ -346,14 +348,9 @@ void HashJoiner::decr_prober(RuntimeState* state) {
     }
 }
 
-size_t HashJoiner::avg_keys_perf_bucket() const {
+float HashJoiner::avg_keys_perf_bucket() const {
```
Comment (trueeyu) [inline be/src/exec/hash_joiner.cpp:351]: per_bucket?

File: be/CMakeLists.txt [hunk 2]
```diff
@@ -541,6 +541,8 @@ if (CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
         # ignore warning from apache-orc
         set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -Wno-unsafe-buffer-usage")
     endif ()
+else ()
+    set(CXX_GCC_FLAGS "${CXX_GCC_FLAGS} -fcoroutines")
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply be/CMakeLists.txt:545]: it will compile failed.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 4 / 4 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|

File: be/src/exec/hash_joiner.cpp (hunk 1 above)
Comment (fzhedu) [reply be/src/exec/hash_joiner.cpp]: fixed

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

File: be/src/exec/hash_joiner.cpp [hunk 3]
```diff
@@ -211,6 +212,7 @@ Status HashJoiner::build_ht(RuntimeState* state) {
         RETURN_IF_ERROR(_hash_join_builder->build(state));
         size_t bucket_size = _hash_join_builder->hash_table().get_bucket_size();
         COUNTER_SET(build_metrics().build_buckets_counter, static_cast<int64_t>(bucket_size));
+        COUNTER_SET(build_metrics().build_keys_per_bucket, static_cast<int64_t>(100 * avg_keys_per_bucket()));
```
Comment (trueeyu) [inline be/src/exec/hash_joiner.cpp:215]: why 100?

File: be/src/exec/hash_joiner.cpp [hunk 4]
```diff
@@ -53,6 +53,7 @@ void HashJoinBuildMetrics::prepare(RuntimeProfile* runtime_profile) {
     build_conjunct_evaluate_timer = ADD_TIMER(runtime_profile, "BuildConjunctEvaluateTime");
     build_buckets_counter = ADD_COUNTER(runtime_profile, "BuildBuckets", TUnit::UNIT);
     runtime_filter_num = ADD_COUNTER(runtime_profile, "RuntimeFilterNum", TUnit::UNIT);
+    build_keys_per_bucket = ADD_COUNTER(runtime_profile, "BuildKeysPerBucket%", TUnit::UNIT);
```
Comment (trueeyu) [inline be/src/exec/hash_joiner.cpp:56]: % ?

Comment (fzhedu) [reply be/src/exec/hash_joiner.cpp:56]: show it as percentage

File: be/src/exec/join_hash_map.tpp [hunk 5]
```diff
@@ -808,63 +811,77 @@ void JoinHashMap<LT, BuildFunc, ProbeFunc>::_search_ht_remain(RuntimeState* stat
...
+        X<first_probe>(state, build_data, data);                                        \
+    }
+
 template <LogicalType LT, class BuildFunc, class ProbeFunc>
 template <bool first_probe>
 void JoinHashMap<LT, BuildFunc, ProbeFunc>::_search_ht_impl(RuntimeState* state, const Buffer<CppType>& build_data,
                                                             const Buffer<CppType>& data) {
+    bool enable_interleaving = state->query_options().interleaving_group_size < 0; // force enable
+    if (state->query_options().interleaving_group_size > 0) {
+        enable_interleaving = ((_table_items->probe_bytes > (1UL << 25) && _table_items->keys_per_bucket > 1.5) ||
```
Comment (trueeyu) [inline be/src/exec/join_hash_map.tpp:836]: why check _table_items->keys_per_bucket > 1 ?

Comment (fzhedu) [reply be/src/exec/join_hash_map.tpp:836]: as tested, if keys_per_bucket is small, the benefit is not obvious, so limit it to get obvious benefit. if keys_per_bucket = 1, each valid bucket just has one key, no need to access next, less random access.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Review by trueeyu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 5 / 84 (05.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hash_joiner.cpp|0|4|00.00%|[56, 215, 351, 353]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|5|00.00%|[66, 67, 69, 70, 72]|
|:large_blue_circle:|src/exec/join_hash_map.h|0|19|00.00%|[131, 133, 134, 135, 136, 138, 139, 196, 197, 199, 200, 201, 206, 207, 209, 246, 247, 248, 250]|
|:large_blue_circle:|src/exec/join_hash_map.tpp|4|53|07.55%|[74, 170, 836, 837, 838, 843, 846, 850, 853, 856, 859, 862, 871, 874, 879, 881, 884, 998, 1000, 1001, 1002, 1003, 1005, 1008, 1009, 1010, 1011, 1013, 1014, 1015, 1017, 1018, 1020, 1021, 1087, 1128, 1237, 1357, 1465, 1536, 1591, 1668, 1748, 1874, 1955, 1987, 2018, 2072, 2077]|
|:large_blue_circle:|src/exec/join_hash_map.cpp|1|3|33.33%|[253, 254]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 5 / 84 (05.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hash_joiner.cpp|0|4|00.00%|[56, 215, 351, 353]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|5|00.00%|[66, 67, 69, 70, 72]|
|:large_blue_circle:|src/exec/join_hash_map.h|0|19|00.00%|[131, 133, 134, 135, 136, 138, 139, 196, 197, 199, 200, 201, 206, 207, 209, 246, 247, 248, 250]|
|:large_blue_circle:|src/exec/join_hash_map.tpp|4|53|07.55%|[74, 170, 836, 837, 838, 843, 846, 850, 853, 856, 859, 862, 871, 874, 879, 881, 884, 998, 1000, 1001, 1002, 1003, 1005, 1008, 1009, 1010, 1011, 1013, 1014, 1015, 1017, 1018, 1020, 1021, 1087, 1128, 1237, 1357, 1465, 1536, 1591, 1668, 1748, 1874, 1955, 1987, 2018, 2072, 2077]|
|:large_blue_circle:|src/exec/join_hash_map.cpp|1|3|33.33%|[253, 254]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 11 / 120 (09.17%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hash_joiner.cpp|0|4|00.00%|[56, 215, 351, 353]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|5|00.00%|[66, 67, 69, 70, 72]|
|:large_blue_circle:|src/exec/join_hash_map.h|0|22|00.00%|[131, 132, 134, 135, 136, 137, 139, 140, 141, 142, 199, 200, 202, 203, 204, 209, 210, 212, 256, 257, 258, 260]|
|:large_blue_circle:|src/exec/join_hash_map.tpp|5|55|09.09%|[74, 170, 287, 294, 341, 773, 845, 848, 852, 855, 858, 861, 864, 873, 876, 881, 883, 886, 1000, 1002, 1003, 1004, 1005, 1007, 1010, 1011, 1012, 1013, 1015, 1016, 1017, 1019, 1020, 1022, 1023, 1089, 1130, 1239, 1359, 1467, 1538, 1593, 1670, 1750, 1876, 1957, 1989, 2020, 2074, 2079]|
|:large_blue_circle:|src/exec/join_hash_map.cpp|6|34|17.65%|[35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 50, 51, 52, 55, 56, 57, 60, 61, 63, 64, 67, 69, 70, 300, 301]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 11 / 120 (09.17%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hash_joiner.cpp|0|4|00.00%|[56, 215, 351, 353]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/hash_join_build_operator.cpp|0|5|00.00%|[66, 67, 69, 70, 72]|
|:large_blue_circle:|src/exec/join_hash_map.h|0|22|00.00%|[131, 132, 134, 135, 136, 137, 139, 140, 141, 142, 199, 200, 202, 203, 204, 209, 210, 212, 256, 257, 258, 260]|
|:large_blue_circle:|src/exec/join_hash_map.tpp|5|55|09.09%|[74, 170, 287, 294, 341, 773, 845, 848, 852, 855, 858, 861, 864, 873, 876, 881, 883, 886, 1000, 1002, 1003, 1004, 1005, 1007, 1010, 1011, 1012, 1013, 1015, 1016, 1017, 1019, 1020, 1022, 1023, 1089, 1130, 1239, 1359, 1467, 1538, 1593, 1670, 1750, 1876, 1957, 1989, 2020, 2074, 2079]|
|:large_blue_circle:|src/exec/join_hash_map.cpp|6|34|17.65%|[35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 50, 51, 52, 55, 56, 57, 60, 61, 63, 64, 67, 69, 70, 300, 301]|

**Review by trueeyu — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Pull request merged**